src/librustc_typeck/check/coercion.rs

```diff
@@ -1011,9 +1044,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let result = self.commit_if_ok(|_| coerce.coerce(new_ty, prev_ty));
             match result {
                 Ok(ok) => {
-                    let (adjustments, target) = self.register_infer_ok_obligations(ok);
-                    self.apply_adjustments(new, adjustments);
-                    return Ok(target);
+                    let coerce = self.register_infer_ok_obligations(ok);
+                    // If mutable reborrow happens, fix up autoderef to use DerefMut
+                    if coerce.reborrow_mut {
+                        self.convert_place_derefs_to_mutable(new);
+                    }
```

Comment:
I'm wondering whether it is possible to figure out `reborrow_mut` just by examining the adjustments? I should think "probably yes". If so, could we remove the code that is computing `reborrow_mut` and instead modify `apply_adjustments` to figure out whether `reborrow_mut` *would have been true* -- and, if so, invoking `self.convert_place_derefs_to_mutable(new)`?

If that is not possible, then I'd still like to see the field moved into the adjustments and have this code be invoked by `apply_adjustments`, instead of having coercion return three things.

Reply:
OK, I just spent a bit of time skimming the code for calls to `apply_adjustments` -- there are quite a few that have no connection to coercion -- but it still *feels* like a better spot to me.

Reply:
But at least based on an *initial* read, this refactoring would likely subsume some of the earlier commits as well, no?

Reply:
I think it's a good idea. I'll go through all callers to see if that's doable.